File: engines/rendering/gecko/src/rendering_context.cpp
/*
 * Nexus Browser - Gecko Rendering Context
 * Copyright (C) 2026
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#include "rendering_context.hpp"

namespace nexus::gecko {

RenderingContext::RenderingContext(
    const EngineConfiguration& configuration
)
    : configuration_(configuration)
{
}

RenderingContext::~RenderingContext()
{
    shutdown();
}

bool RenderingContext::initialize()
{
    if (initialized_) {
        return true;
    }

    initialized_ = true;

    return true;
}

bool RenderingContext::shutdown()
{
    initialized_ = false;

    return true;
}

bool RenderingContext::render()
{
    if (!initialized_) {
        return false;
    }

    ++metrics_.frames_rendered;

    return true;
}

bool RenderingContext::execute_javascript(
    const std::string& script
)
{
    if (!initialized_) {
        return false;
    }

    if (!configuration_.javascript_enabled) {
        return false;
    }

    if (script.empty()) {
        return true;
    }

    ++metrics_.scripts_executed;

    return true;
}

bool RenderingContext::set_viewport(
    std::uint32_t width,
    std::uint32_t height
)
{
    if (width == 0 || height == 0) {
        return false;
    }

    viewport_width_ = width;
    viewport_height_ = height;

    return true;
}

std::uint32_t
RenderingContext::viewport_width() const
{
    return viewport_width_;
}

std::uint32_t
RenderingContext::viewport_height() const
{
    return viewport_height_;
}

bool RenderingContext::is_initialized() const
{
    return initialized_;
}

const RenderingMetrics&
RenderingContext::metrics() const
{
    return metrics_;
}

} // namespace nexus::gecko

File: engines/rendering/servo/src/rendering_context.rs
/*
 * Nexus Browser - Servo Rendering Context
 * Copyright (C) 2026
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

use crate::pipeline::{RenderingPipeline, Viewport};

pub struct RenderingContext {
    pipeline: RenderingPipeline,
    javascript_enabled: bool,
}

impl RenderingContext {
    pub fn new(javascript_enabled: bool) -> Self {
        Self {
            pipeline: RenderingPipeline::new(),
            javascript_enabled,
        }
    }

    pub fn initialize(&mut self) -> Result<(), String> {
        self.pipeline.initialize()
    }

    pub fn shutdown(&mut self) -> Result<(), String> {
        self.pipeline.shutdown()
    }

    pub fn render(&mut self) -> Result<(), String> {
        self.pipeline.render()
    }

    pub fn execute_javascript(
        &mut self,
        script: &str,
    ) -> Result<(), String> {
        if !self.javascript_enabled {
            return Err(
                "JavaScript is disabled".to_string()
            );
        }

        if !self.pipeline.is_running() {
            return Err(
                "Rendering pipeline is not running"
                    .to_string()
            );
        }

        if !script.is_empty() {
            self.pipeline.record_script_execution();
        }

        Ok(())
    }

    pub fn set_viewport(
        &mut self,
        width: u32,
        height: u32,
    ) -> Result<(), String> {
        self.pipeline.set_viewport(width, height)
    }

    pub fn viewport(&self) -> Viewport {
        self.pipeline.viewport()
    }

    pub fn is_initialized(&self) -> bool {
        self.pipeline.is_running()
    }
}
